### ISSUE TITLE
Question: Is there an option to only output on Failed hook

### ISSUE THREAD

**Issue by graemedavidson** [pre-commit/pre-commit#823]
Is it possible to have it only output on Failure, otherwise leaving it clear.

Might be a good additional option (if it does not exist already) to add at the parent level a verbosity setting in which you can get the current Passed/Skipped/Failed output at the highest. a mid level providing a single line of stats --passed=5 --skipped=2 and the above with only failures.

**Comment by asottile** [comment]
That's the default behaviour unless you're using `--verbose` or `verbose: true`

**Comment by asottile** [comment]
no, it should always show the passed messaging

**Comment by graemedavidson** [comment]
Ok, so my question becomes a feature request.

On 5 Sep 2018 18:56, "Anthony Sottile" <notifications@github.com> wrote:

no, it should always show the passed messaging

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pre-commit/pre-commit/issues/823#issuecomment-418822945>,
or mute the thread

**Comment by asottile** [comment]
What's the motivation? My initial lean is -0

**Comment by asottile** [comment]
Unless you have a good reason I don't want to or plan on implementing this.

Comment back and I'll reopen though :)

**Comment by aschokking** [comment]
I was looking for an option like this (silence the passed hooks) so that there's less noise in my output to scroll through.

**Comment by asottile** [comment]
CC @blueyed

Even if a commandline option were added (which I'm against -- the output is almost a _branding_ at this point), there wouldn't be a way to set it since there aren't ways to pass through arguments from `git`

**Comment by asottile** [comment]
oops, meant to reopen -- should be open now. let's discuss.

**Comment by blueyed** [comment]
I would (like to) configure this in the config file.

**Comment by blueyed** [comment]
So I assume there is no user config then?
What about an environment variable?

**Comment by blueyed** [comment]
A git config option could be used for this.

**Comment by asottile** [comment]
I guess another option would be to set it during `pre-commit install` I forgot about this but we already have a precedent for an option there: https://github.com/pre-commit/pre-commit/blob/518a72d7e7598b71105fb764d990438c1b6d485a/pre_commit/resources/hook-tmpl#L20

**Comment by blueyed** [comment]
@asottile 
Also sounds good.

I came up with the following for me:
```diff
--- .git/hooks/pre-commit	2019-03-02 15:25:46.671907241 +0100
+++ .git/hooks/pre-commit.new	2019-03-02 15:25:35.715359949 +0100
@@ -162,8 +162,23 @@
 def main():
     retv, stdin = _run_legacy()
     try:
+        print('Running pre-commit: ', end='', flush=True)
         _validate_config()
-        return retv | subprocess.call(_exe() + _opts(stdin))
+        ret = subprocess.run(_exe() + _opts(stdin) + ('--color=always',),
+                             stdout=subprocess.PIPE)
+        out = ret.stdout.decode()
+        if ret.returncode == 0:
+            lines = out.splitlines()
+            for line in lines:
+                if line.endswith('Passed\x1b[0m'):
+                    print('✔', end='')
+                elif line.endswith('Skipped\x1b[0m'):
+                    print('…', end='')
+            print()
+        else:
+            print('\r', end='')
+            print(out)
+        return retv | ret.returncode
     except EarlyExit:
         return retv
     except FatalError as e:
```

**Comment by asottile** [comment]
I meant more that you'd have a `QUIET` option there which forwards along `--quiet`

**Comment by rbu** [comment]
I'm a big fan of this, too. Most developers in my team only care about the checks that failed and not those that pass, especially if they are skipped. Can I advocate for a repo-level default for "quiet" (in .pre-commit-config.yaml) and an option (via environment variable?) to override that per user if needed?

**Comment by blueyed** [comment]
@asottile 
My solution above was really only a POC to get this going for me with pytest, and it is working well since then.

**Comment by allburov** [comment]
Hey there!
We have a big mono repository with 30 components and every component has his own hooks: `mypy`, `black` and `reorder-python-imports`. We need all of them, because configuration may be different (legacy code base and other causes)

So, when I've changed a one file, have made some mistakes and commit that, wonderful (it's really great tool, thank you!) tool `pre-commit` get me 89 lines with `Skipped` and only one with `Failed` status **in the middle** of this output.

It's really hard to parse these lines, we must use mouse to scroll to `Failed` hook. Also when PyCharm or GitKraken is used to commit, they has a small popup, like this:

![image](https://user-images.githubusercontent.com/4376814/66737964-8926bb00-ee97-11e9-9f1a-302117161017.png)

I just want to see the problem or "All is ok" otherwise.

How we can solve this problem? I suggest this way:

1. **MUST** Add ability to pass `--show: [skipped,failed,passed]` to command line interface. 
2. **MUST** Use env variables `PRE_COMMIT_SHOW="failed, passed" pre-commit run`
3. (may be, need to discussion) Add config option to `.pre-commit-config.yaml`.
```yaml
show:
  - skipped
  - failed
  - passed
```
>that would change it for all users which I wouldn't want so I don't think that's an appropriate place to change this

But `pre-commit` already has `fail_fast` parameter which, as for me, must not apply for all users too. But that's applied, so why not use `show` also in config?
4. (may be, need to discussion) Add summary information below output: 
```
... a lot of SKIPPED hook
hookname ..... PASSED
hookname ..... FAILED
hookname ..... SKIPPED
hookname ..... SKIPPED
... a lot of SKIPPED hook

Hooks result:
Total: 13 
Passed: 0
Failed: 1
Skipped: 12
```
for `--show=failed,passed` it will be
```
hookname ..... PASSED
hookname ..... FAILED

Hooks result:
Total: 13 
Passed: 0
Failed: 1
Skipped: 12
```

**pre-commit install --quiet** isn't good idea, it looks like "You need uninstall\install your software to apply new config". My first reaction is "Oh really, why we could change config in place?"

@asottile If you don't mind for that implementation, I can contribute 1\2 things from my suggestion - supports cli and variables. It will be enough for me and my use case :)

**Comment by allburov** [comment]
@asottile 
Or we can fully change output, copy it from pytest. @blueyed has show us PoC
```bash
pre-commit run # or git commit -m 'Done IS-111'
# . means PASSED
# s means SKIPPED
# F means FAILED
Hooks: ..........Fsssssssssss


# Show a error output only after all
flake8/ptaf-conf-mgr.....................................................Failed
hookid: flake8

services/applications.py:17:1: I100 Import statements are in the wrong order. 'from project.tenants import is_read_only_tenant' should be before 'from project.errors import OperationError'
```

If we will choose this way, we don't need `show` option at all. 
I wuold prefer this way :)

**Comment by asottile** [comment]
I do not want to drastically change the brand look, so anything like pytest is out.

The most extreme I think I'd be ok with is something like:

```
Running 89 hooks..............................................................................Passed
```

vs

```
Running 89 hooks...................................................................................Failed
flake8.............................................................................................................Failed
hookid: flake8

...

```

but this needs a concrete proposal (configuration value, schema, output scheme, etc.)

**Comment by asottile** [comment]
@allburov I don't think the proposed `hide_skipped` is workable -- it's now impossible to discern the difference between "pre-commit didn't run" and "everything was skipped"

**Comment by allburov** [comment]
@asottile you take a words from my mouth. 
May be we should add `collapse_skipped: True|False` (or `combine_skipped`)? And add summary information at the bottom (with the brand look).
```
mypy......................................................Passed
flake8.....................................................Failed
hookid: flake8

....

97 of 98 hooks.....................................Skipped
```

If "everything was skipped" it will be single line:
```bash
All hooks.............................................Skipped
# or
98 of 98 hooks....................................Skipped
```

With `pre-commit run -v` output will be full.

**Comment by asottile** [comment]
I don't think that addresses everyone's desires (and opens the door to 3-4 more similar options which I definitely don't want)

**Comment by allburov** [comment]
Do you mean we need go back to disscussion with "universal" solution `show_statuses: failed,skipped,passed`?

**Comment by asottile** [comment]
no I think we need something simpler like `quiet: true`, maybe something like what I've shown above where there's a single line and then failures if present?

**Comment by allburov** [comment]
But how we could show "the progress"? We could use "dot" for show progress, but if we will have 100 hooks - it can occupy two or more lines in output, it's not the brand look.

>opens the door to 3-4 more similar options which I definitely don't want

This solution does not resolved this, because someone can ask "Get me passing hooks in quite mode too!"

We should show the progress, of course https://github.com/pre-commit/pre-commit/blob/master/pre_commit/commands/run.py#L111-L112

**Comment by asottile** [comment]
we already don't show progress, but it's planned in the future 🤷‍♂

**Comment by allburov** [comment]
w\a to hide all skipping msg from output

[0001-hide-skipped-hook-from-output.txt](https://github.com/pre-commit/pre-commit/files/3732422/0001-hide-skipped-hook-from-output.txt)
```bash
cd your-project
wget https://github.com/pre-commit/pre-commit/files/3732422/0001-hide-skipped-hook-from-output.txt
# if you use venv directory or specify you path to site-packages
patch -p1 -d venv/lib/python3.7/site-packages/ < ../pre-commit/0001-hide-skipped-hook-from-output.txt
```

**Comment by allburov** [comment]
>we already don't show progress, but it's planned in the future

I mean pre-commit show that "Current hook is running, wait". With `quite` it will look like (with 90 hooks) "All hooks is running wait ... forever?"

**Comment by allburov** [comment]
I just want to hide a lot of "Skipped" hooks and have useful output - what hooks are passed or failed.

**Comment by asottile** [comment]
I think the argument by others in the thread is that even the Passed ones aren't particularly useful either

**Comment by blueyed** [comment]
You can use a spinner to display progress with many tasks.
FWIW https://github.com/pre-commit/pre-commit/issues/823#issuecomment-468925217 still works great for me (used with pytest, which does not have many hooks).

**Comment by asottile** [comment]
I will not add a spinner and we've already written off that patch

**Comment by blueyed** [comment]
@asottile 
Well, it's not clear to me otherwise I would have not asked.

Because of the "brand look"?

> I do not want to drastically change the brand look, so anything like pytest is out.

(pytest is not using a spinner as you know)

Anyway, I'm out of here - my workaround is OK for me with pytest.. :man_shrugging:

**Comment by allburov** [comment]
@asottile I've reviewed again your solution with "quiet" mode it's good solution. With "quiet" mode we don't have to show progress at all because it's "quiet" mode.

Could we summarize what do you expect?
1. Add flag `--quiet, -q` to command `pre-commit run`, which made output as you suggested above https://github.com/pre-commit/pre-commit/issues/823#issuecomment-541705583
```bash
> pre-commit run -q
Running 89 hooks...........................................   # We've run, print welcome message and are waiting all hooks will pass
# With all Passed or all Skipped hooks it will be:
Running 89 hooks........................................... Passed
Running 89 hooks........................................... Skipped 

# If any hook failed, print Failed an hook's output.
Running 89 hooks........................................... Failed

flake8.....................................................Failed
hookid: flake8

...
... git diff 
...
```
1.1. Introduce env variable  `PRE_COMMIT_QUIET=1` inside `pre-commit run`, which allow users to enable quiet mode by default in their PC. 
2. Add install option `--output-mode=[normal|quite]` (with `normal` by default), which add to `hook-tmp` option `PRE_COMMIT_QUIET = None # True` and run `pre-commit run `  with or without `--quiet` flag. 
2. *alternative* We can add `addopts` flag to `pre-commit install` command, which allow user to add any option (for the future needed). So, I will need run this command to install pre-commit in quiet mode
```bash
pre-commit install --addopts="--quite"
# But it's potential point for mistakes and typos
```

I've modified your suggestion in the topic with "`pre-commit install --quiet`", because it makes a little confuse - it looks like we run install process with "quiet" mode, not "install and configure always run with quiet flag"

**Comment by asottile** [comment]
seems fine, need to figure out how `verbose` and `quiet` interact, let's skip the environment variable for the first implementation,  let's not do `--addopts` it's going to be a footgun (what if someone did `--addopts=--files` 💥)

**Comment by kolinkorr839** [comment]
Just curious what is the workaround for this? I am using version 2.9.3. Is there a patch that I can apply so that I only 'failed' hooks?

**Comment by peterjc** [comment]
@kolinkorr839 you could try my branch on #1560, not looked at this lately to see if there is anything else more recent in progress.

**Comment by lucasteligioridis** [comment]
Is a quiet mode an option for this at all? I really do find the output with noise I simply don't care about, I only really care about failures.

It's been over a year since there has been any activity on this issue, I honestly think the option described here is good enough and will eliminate so much noise from the output:
https://github.com/pre-commit/pre-commit/issues/823#issuecomment-541705583

**Comment by mustafa0x** [comment]
Anyway to hide **Skipped** checks?

**Comment by asottile** [comment]
given the recent surge of ai slop prs "addressing" this I'm just going to close this out -- I don't really intend to add such a feature and it doesn't really seem "simple" to address everyone's needs here
